### PR TITLE
interrupt_controller: gic: Support PPIs

### PR DESCRIPTION
**Pull request by carlocaione**
The GIC-400 driver currently only supports SPIs because the (32) offset
for the INTIDs is hard-coded in the driver. At the driver level there is
no really difference between PPIs and SPIs so we can easily extend the
driver to support PPIs as well.

This is useful if we want to add support for the ARM Generic Timers that
use INTIDs in the PPI range.

SPI interrupts are in the range [0-987]. PPI interrupts are in the
range [0-15].

This patches introduces a new interrupt type (GIC_SPI | GIC_PPI) to be
defined in the DTS to mark the IRQ as SPI or PPI, following what's done
in the Linux kernel. When the IRQ is marked as SPI its INTID is
offsetted by 32, otherwise by 16 when marked as PPI. The fix is done
introducing a new macro to be used in the DTS fixup file.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

### CODE REVIEW

**Comment by carlocaione** [comment]
Gentle ping tagging @bbolen @stephanosio @ioannisg @andrewboie because previously involved with related stuff in the past.

File: include/dt-bindings/interrupt-controller/arm-gic.h
Comment (stephanosio) [inline include/dt-bindings/interrupt-controller/arm-gic.h]: I think this should provide a cleaner and much easier to use interface:
```
#define GIC_PPI_BASE 16
#define GIC_SPI_BASE 32

#define GIC_PPI(id) (GIC_PPI_BASE + id)
#define GIC_SPI(id) (GIC_SPI_BASE + id)
```

With this, in dts, we can just do:
```
interrupts = <GIC_SPI(36) IRQ_DEFAULT_PRIORITY
		IRQ_TYPE_LEVEL>, ...
```

Also we don't need dts_fixup for every GIC mapped interrupt this way.

Comment (carlocaione) [reply include/dt-bindings/interrupt-controller/arm-gic.h]: Yes. I chose that way to be closer to how that is done in the Linux kernel but I like the fact that we can avoid the dts_fixup.

**Comment by stephanosio** [comment]
Also it may be worth noting that there was some major refactoring done on GIC driver and Cortex-R interrupt system in general here: https://github.com/zephyrproject-rtos/zephyr/pull/19698/commits/5772e5245b457809cc922d08a99486391c5f9499 https://github.com/zephyrproject-rtos/zephyr/pull/19698/commits/82dd67b32f05ca121fff17cbcb09f0931196cbef

**Comment by carlocaione** [comment]
@stephanosio damn, I was not aware of that. At this point I think the best way forward is just for you to incorporate these changes. Is that ok for you?

**Comment by stephanosio** [comment]
> @stephanosio damn, I was not aware of that. At this point I think the best way forward is just for you to incorporate these changes. Is that ok for you?

Since we're already here (and #19698 has already become way too big for what it was initially intended for), I think it would be better if these changes get reviewed and merged here.

It is very likely that this PR will get merged before #19698, so once that happens, it will cause a merge conflict for #19698 and I will make the necessary adoptions there to incorporate these changes.

**Comment by carlocaione** [comment]
@stephanosio ACK. Pushed a new rev.

**Comment by galak** [comment]
Since you are reworking this can we get the dts for GIC to align with https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/interrupt-controller/arm,gic.yaml

**Comment by stephanosio** [comment]
Yea, the current Zephyr interrupt API won't be able to accommodate the Linux GIC dts without introducing many hacks (such as specifying dts_fixup for every GIC mapped interrupt, which is a major hassle).

The main reason is that Linux IRQ control API accepts `struct irq_data`, whereas Zephyr only accepts single `unsigned int irq`

**Comment by carlocaione** [comment]
agree. @galak given that is unlikely that we can align with the Linux kernel soon, I propose to merge the commit as is (with yours and @stephanosio ACK).

**Review by stephanosio — APPROVED** [review]
NOTE: Once merged, to be re-integrated into the refactored gic driver in #19698; unit tests for in-depth functional verification is to be implemented in the future.

**Review by bbolen — APPROVED** [review]

**Comment by stephanosio** [comment]
cc @carlocaione @timork @vladimir-podbrezsky

**Comment by carlocaione** [comment]
@galak can you merge this? thanks.

**Comment by carlocaione** [comment]
@galak please, consider this blocked on #20243. So wait for that before merging this.

**Comment by stephanosio** [comment]
> What the issue w/regards to dts_fixup.

For GIC, there are multiple interrupt types with different IRQ bases. If we follow the Linux dts and specify interrupt type in a separate property, for every single IRQ, we have to map/fix-up, the complex (interrupt type)+(type-dependent interrupt index) definition from dts to the actual linear IRQ number used by the Zephyr interrupt API.

Refer to the following for what has to be done if we follow the Linux dts:
https://github.com/zephyrproject-rtos/zephyr/commit/b761bd4ca3580b454ed5c25ea94ac2907ea2e3f3#diff-8dec1584212c37fe3f73aa7c195ebde2L9-L16

This is simply not practical. If we really wanted to follow the Linux dts, we will have to redesign the Zephyr interrupt API to either accept interrupt type parameter alongside the IRQ number (which is somewhat interrupt controller-specific and should be avoided) or some kind struct representing an IRQ (as Linux is doing, but this is really an overkill for an OS that is intended to run on environments with very limited resources).

IMHO, following Linux dts is good while it makes sense; but, if you have to sacrifice tidiness of code and sanity of developers over following Linux dts, that is a very questionable approach.

**Comment by stephanosio** [comment]
Please merge this if there is no further comment. I have a quite few changes that are dependent on this.

**Comment by stephanosio** [comment]
Ping

**Comment by galak** [comment]
> Ping

Sorry was busy with ELC-E last week.  I'd like to see if @ulfalizer has any ideas on how to handle the IRQ numbering if we try and keep a bit more in sync with the linux binding.

**Review by galak — CHANGES_REQUESTED** [review]
So I think at a minumum we can tweak the binding to a bit more in-sync with linux while leaving the IRQ encoded as Zephyr would desire:

```
interrupt-cells:
  - irq-type
  - irq
  - flags
  - priority
```

So the interrupts would look something like:

```
interrupts = <GIC_SPI GIC_SPI_IRQ(42) IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY ...
```

We'd have in the header:
```
#define GIC_SPI 0
#define GIC_PPI 1
#define GIC_SPI_IRQ(id) (GIC_SPI_BASE + id)
#define GIC_PPI_IRQ(id) (GIC_PPI_BASE + id)
```

**Comment by stephanosio** [comment]
> So I think at a minumum we can tweak the binding to a bit more in-sync with linux while leaving the IRQ encoded as Zephyr would desire

@galak @ulfalizer I guess that would be one way to do it if we are really desperate about aligning with Linux dts (to be honest, I don't see the point of having `irq-type` which is redundant and does nothing whatsoever at this time).

Maybe if we add some func to the dt header generation script that, for GIC, automatically interprets `irq-type` to add type-specific base to `irq` for the resulting IRQ symbol in dt header, it would make more sense; but that sounds like a whole other issue.

**Comment by carlocaione** [comment]
> So I think at a minumum we can tweak the binding to a bit more in-sync with linux while leaving the IRQ encoded as Zephyr would desire:

I found this not elegant and it definitely increases the error surface (you are defining twice the IRQ type with no real gain just for the sake to be closer to what's done on Linux).

**Comment by stephanosio** [comment]
> So I think at a minumum we can tweak the binding to a bit more in-sync with linux while leaving the IRQ encoded as Zephyr would desire

@galak Could we do that when we add an automation script for IRQ type-specific number handling in the future? As carlocaione already pointed out, this could be an error-prone approach (having a redundant and non-functional cell).

Changes to the user dts will be required either way, and changing from 3 to 4 cells will actually ensure nothing fails silently (if we keep the number of cells same across the change, if a user decides to not use `GIC_*_IRQ()` macro and specifies a zero-indexed IRQ number instead at the moment, no compilation time error will be generated after the automation script is introduced and `irq` cell will have an invalid value without user noticing).

**Comment by galak** [comment]
> @galak Could we do that when we add an automation script for IRQ type-specific number handling in the future? As carlocaione already pointed out, this could be an error-prone approach (having a redundant and non-functional cell).

Chatted briefly with @ulfalizer and thought was we could do something in scripts/dts/gen_defines.py to special case create a Zephyr IRQ value for 'arm,gic' (we already does this for multi-level IRQs) [see the `write_irqs` function]
 
> Changes to the user dts will be required either way, and changing from 3 to 4 cells will actually ensure nothing fails silently (if we keep the number of cells same across the change, if a user decides to not use `GIC_*_IRQ()` macro and specifies a zero-indexed IRQ number instead at the moment, no compilation time error will be generated after the automation script is introduced and `irq` cell will have an invalid value without user noticing).

My main concern is alignment as I expect that as Cortex-A support goes increases the desire to utilize a device tree from linux directly is only going to increase.  If we want to change the GIC binding for our needs we should work with the linux guys to define/update the binding to match.  DTSs are meant to be agnostic of software, so we should not have a zephyr specific focus.

I'd actually prefer we drop priority from the binding as well.

**Comment by stephanosio** [comment]
> I expect that as Cortex-A support goes increases the desire to utilize a device tree from linux directly is only going to increase.

> DTSs are meant to be agnostic of software

That certainly is a valid point, but I wonder how far we can manage before getting blocked by some fundamentally irreconcilable differences between Linux and Zephyr (given that neither Linux or Zephyr dts is truly agnostic of implementation details)

Have you assessed the feasibility of this previously somewhere (maybe in another issue)? I think this could be a discussion worth hundreds of pages in writing.

**Comment by stephanosio** [comment]
@carlocaione @galak @ulfalizer I implemented the aforementioned automatic fix-up script and changed the dt binding to have separate `type` and `irq` (type-specific index, rather than a linear one) cells, just like Linux.

Please review and merge this if there is no problem with it.

**Comment by carlocaione** [comment]
Thank you for this. LGTM.

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: Could check if `"type"` exists in the dictionary and call `err()` otherwise (see elsewhere in the script). Maybe `map_arm_gic_irq_type()` could take just `irq`, to move the check into it.

Probably will never happen, but the goal is for the script to never crash with a generic Python error.

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: Could do early returns maybe:

```python
def map_arm_gic_irq_type(irq, irq_type, irq_num):
    # Maps ARM GIC IRQ (type)+(index) combo to linear IRQ number
    if irq_type == 0: # GIC_SPI
        return irq_num + 32
    if irq_type == 1: # GIC_PPI
        return irq_num + 16
    return irq_num
```

Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Python nit: Could put two spaces before `#` after code, for consistency. It's a common style in Python.

**Comment by ulfalizer** [comment]
I haven't kept up with the details here, but fwiw, I think we should ignore what Linux does if there's something simpler we could do.

Linux stuff is generally of pretty high quality, but not everything from there is super solid, so we shouldn't unconditionally worship it. The Kconfig code is pretty horrendous for example, though I have a lot of respect for the design of Kconfig. :)

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: Unused `irq` argument.

**Review by ulfalizer — CHANGES_REQUESTED** [review]
(For unused argument. See comments too.)

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: BTW, if you think it gets clearer, feel free to move both this function and `encode_zephyr_multi_level_irq()` (and maybe `irq_name_alias()`) out of `write_irqs()` and putting them after it (I usually put top-level code at the top).

Know it was just for consistency, but `write_irqs()` might be getting a bit unwieldy. Unrelated to this PR though...

File: scripts/dts/gen_defines.py
Comment (stephanosio) [reply scripts/dts/gen_defines.py]: Added check for `type` definition as well as its value.

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: Could align it like this just for consistency:

```python
err("Expected binding for {!r} to have 'type' in *-cells"
    .format(irq.controller))
```

File: scripts/dts/gen_defines.py
Comment (stephanosio) [reply scripts/dts/gen_defines.py]: > BTW, if you think it gets clearer, feel free to move both this function and encode_zephyr_multi_level_irq() (and maybe irq_name_alias()) out of write_irqs() and putting them after it (I usually put top-level code at the top).

I am not really a Python guy, so I will leave it to someone else who is more acquainted with the language.

**Review by ulfalizer — APPROVED** [review]

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Could maybe say `interrupt-cells` instead of `*-cells`, since it's known here.

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Yeah, no worries. Would've been a bit silly to force into this PR anyway.

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [reply scripts/dts/gen_defines.py]: Thanks!

File: scripts/dts/gen_defines.py
Comment (stephanosio) [reply scripts/dts/gen_defines.py]: Fixed.

File: scripts/dts/gen_defines.py
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: Ah... had missed this one. :)

File: dts/bindings/interrupt-controller/arm,gic.yaml
Comment (galak) [inline dts/bindings/interrupt-controller/arm,gic.yaml]: Can we swap `-flags` and `-priority` order.

**Review by galak — CHANGES_REQUESTED** [review]
Looks good, one change I'd like to see is swapping the order of `flags` and `priority` in the interrupt-cells.  This keeps is a bit more compatible with the linux binding, and pretty minor change at this point before we have more dts files that have GICs in them :)

File: dts/bindings/interrupt-controller/arm,gic.yaml
Comment (stephanosio) [reply dts/bindings/interrupt-controller/arm,gic.yaml]: Sounds fair.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Thanks, I know this was a bit long to get ready.

**Review by bbolen — COMMENTED** [review]
Sorry, I'm trying to catch up on all of the cortex changes.
I might suggest squashing this commit on the first one.

**Comment by carlocaione** [comment]
Squashed the commit as suggested by @bbolen and force-pushed again. We should actually squash all the commits together if we care about bisectability.

FWIW #20263 depends on this story.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]
Looks good to me

**Pull request merged**
